src/util.js

```diff
@@ -60,16 +97,15 @@ function serviceResponse(service, stages, finalizer) {
       enumerable: false, writable: false, configurable: false,
       value: (resolve, reject) => {
         let result = client.executePipeline(Array.isArray(stages) ? stages : [ stages ]);
-        return (!!finalizer) ?
-          result.then(finalizer).then(resolve, reject) : result.then(resolve, reject);
+        return (!!finalizer) ? result.then(collectMetadata(finalizer)).then(resolve, reject) : result.then(resolve, reject);
```

Comment:
The use of `finalizer`s seems fishy and makes the core logic of `collectMetadata` harder to reason about (checking if we've already attached once). It seems like these are necessary due to the syntactic sugar we have. Might it make sense to have execute pipeline have the `finalizer` logic instead to avoid having to use `collectMetadata` here? Either that or maybe `executePipeline` is really `serviceResponse`.